File: Intermediate/33.Write a program to detect a cycle in a directed graph using DFS/Solution.py
def dfs_cycle(v, graph, visited, rec_stack):
    # Mark current node as visited and add it to recursion stack
    visited[v] = True
    rec_stack[v] = True

    # Visit all neighbors
    for neighbor in graph[v]:
        if not visited[neighbor]:
            if dfs_cycle(neighbor, graph, visited, rec_stack):
                return True
        elif rec_stack[neighbor]:
            return True

    # Remove the vertex from recursion stack before returning
    rec_stack[v] = False
    return False

def detect_cycle(graph, V):
    visited = [False] * V
    rec_stack = [False] * V

    for i in range(V):
        if not visited[i]:
            if dfs_cycle(i, graph, visited, rec_stack):
                return True
    return False

def main():
    V = int(input("Enter the number of vertices: "))
    E = int(input("Enter the number of edges: "))

    # Initialize the graph as an adjacency list
    graph = [[] for _ in range(V)]
    print("Enter each edge (u v) [directed edge from u to v]:")
    for _ in range(E):
        u, v = map(int, input().split())
        graph[u].append(v)
    
    if detect_cycle(graph, V):
        print("Cycle detected in the graph.")
    else:
        print("No cycle detected in the graph.")

if __name__ == "__main__":
    main()

File: Intermediate/33.Write a program to detect a cycle in a directed graph using DFS/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

// DFS function to detect cycle in a directed graph
bool dfsCycle(int v, const vector<vector<int>>& graph, vector<bool>& visited, vector<bool>& recStack) {
    visited[v] = true;
    recStack[v] = true;

    // Explore all neighbors
    for (int neighbor : graph[v]) {
        // If neighbor is not visited, do DFS on it
        if (!visited[neighbor]) {
            if (dfsCycle(neighbor, graph, visited, recStack))
                return true;
        }
        // If neighbor is in the recursion stack, cycle detected
        else if (recStack[neighbor]) {
            return true;
        }
    }

    // Remove vertex from recursion stack before returning
    recStack[v] = false;
    return false;
}

// Function to detect cycle in the entire graph
bool detectCycle(const vector<vector<int>>& graph, int V) {
    vector<bool> visited(V, false);
    vector<bool> recStack(V, false);

    for (int i = 0; i < V; i++) {
        if (!visited[i]) {
            if (dfsCycle(i, graph, visited, recStack))
                return true;
        }
    }
    return false;
}

int main() {
    int V, E;
    cout << "Enter the number of vertices: ";
    cin >> V;
    cout << "Enter the number of edges: ";
    cin >> E;

    // Initialize graph as an adjacency list (0-indexed vertices)
    vector<vector<int>> graph(V);
    cout << "Enter each edge (u v) [directed edge from u to v]:" << endl;
    for (int i = 0; i < E; i++) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
    }

    if (detectCycle(graph, V))
        cout << "Cycle detected in the graph." << endl;
    else
        cout << "No cycle detected in the graph." << endl;

    return 0;
}
